src/linear_algebra/bilinear_form.lean

```diff
@@ -68,125 +67,177 @@ rfl
-      rw [bilin_smul_right, ←mul_assoc, mul_comm c, mul_assoc]} },
-  smul_add := λ c B D, by {ext, unfold coe_fn has_coe_to_fun.coe bilin, rw left_distrib},
-  add_smul := λ c B D, by {ext, unfold coe_fn has_coe_to_fun.coe bilin, rw right_distrib},
-  mul_smul := λ a c D, by {ext, unfold coe_fn has_coe_to_fun.coe bilin, rw mul_assoc},
-  one_smul := λ B, by {ext, unfold coe_fn has_coe_to_fun.coe bilin, rw one_mul},
-  zero_smul := λ B, by {ext, unfold coe_fn has_coe_to_fun.coe bilin, rw zero_mul},
-  smul_zero := λ B, by {ext, unfold coe_fn has_coe_to_fun.coe bilin, rw mul_zero} }
-
-lemma smul_apply (a : R₂) (x y : M) : (a • F) x y = a • (F x y) := rfl
-
-/-- `B.to_linear_map` applies B on the left argument, then the right.  -/
-def to_linear_map : M →ₗ[R₂] M →ₗ[R₂] R₂ :=
-linear_map.mk₂ R₂ F.1 (bilin_add_left F) (bilin_smul_left F) (bilin_add_right F) (bilin_smul_right F)
-
-/-- Bilinear forms are equivalent to maps with two arguments that is linear in both. -/
-def bilin_linear_map_equiv : (bilin_form R₂ M) ≃ₗ[R₂] (M →ₗ[R₂] M →ₗ[R₂] R₂) :=
-{ to_fun := to_linear_map,
+      by { unfold coe_fn has_coe_to_fun.coe bilin, rw [bilin_add_left, left_distrib] },
+    bilin_smul_left := λ a x y, by { unfold coe_fn has_coe_to_fun.coe bilin,
+      rw [bilin_smul_left, ←mul_assoc, mul_comm c, mul_assoc] },
+    bilin_add_right := λ x y z, by { unfold coe_fn has_coe_to_fun.coe bilin,
+      rw [bilin_add_right, left_distrib] },
+    bilin_smul_right := λ a x y, by { unfold coe_fn has_coe_to_fun.coe bilin,
+      rw [bilin_smul_right, ←mul_assoc, mul_comm c, mul_assoc] } },
+  smul_add := λ c B D, by { ext, unfold coe_fn has_coe_to_fun.coe bilin, rw left_distrib },
+  add_smul := λ c B D, by { ext, unfold coe_fn has_coe_to_fun.coe bilin, rw right_distrib },
+  mul_smul := λ a c D, by { ext, unfold coe_fn has_coe_to_fun.coe bilin, rw mul_assoc },
+  one_smul := λ B, by { ext, unfold coe_fn has_coe_to_fun.coe bilin, rw one_mul },
+  zero_smul := λ B, by { ext, unfold coe_fn has_coe_to_fun.coe bilin, rw zero_mul },
+  smul_zero := λ B, by { ext, unfold coe_fn has_coe_to_fun.coe bilin, rw mul_zero } }
+
+@[simp]
+lemma smul_apply {R : Type*} [comm_semiring R] [semimodule R M]
+  (B : bilin_form R M) (a : R) (x y : M) :
+  (a • B) x y = a • (B x y) :=
+rfl
+
+end
+
+end bilin_form
+
+section equiv_lin
+
+/-- A map with two arguments that is linear in both is a bilinear form.
+
+This is an auxiliary definition for the full linear equivalence `linear_map.to_bilin`.
+-/
+def linear_map.to_bilin_aux (f : M₂ →ₗ[R₂] M₂ →ₗ[R₂] R₂) : bilin_form R₂ M₂ :=
+{ bilin := λ x y, f x y,
+  bilin_add_left := λ x y z, (linear_map.map_add f x y).symm ▸ linear_map.add_apply (f x) (f y) z,
+  bilin_smul_left := λ a x y, by rw [linear_map.map_smul, linear_map.smul_apply, smul_eq_mul],
+  bilin_add_right := λ x y z, linear_map.map_add (f x) y z,
+  bilin_smul_right := λ a x y, linear_map.map_smul (f x) a y }
```

Comment:
Any reason not to inline this below?

Reply:
`to_bilin` is already quite long, so I would like to keep them separate.